### PR TITLE
feat: Add msgpack support to Realtime

### PR DESCRIPTION
**Pull request by garrensmith**
This is ongoing work on the Realtime functionality. This adds support to encode and decode messages that are encoded with msgpack.  We also didn't correctly handle attach/detach messages. This has been fixed.

### CODE REVIEW

File: server/services/v1/realtime/device_session.go [hunk 1]
```diff
@@ -146,6 +147,7 @@ func (session *Session) Start(ctx context.Context) error {
 		}
 		session.lastReceived = time.Now()
 		if errEvent := session.onMessage(ctx, message); errEvent != nil {
+			log.Info().Msgf("sending error", errEvent)
```
Comment (pboros) [inline server/services/v1/realtime/device_session.go:150]: How verbose can this be? Should it be a metric instead?

**Review by pboros — APPROVED** [review]

File: server/services/v1/realtime/device_session.go (hunk 1 above)
Comment (garrensmith) [reply server/services/v1/realtime/device_session.go:150]: I think we will add metrics in a later PR

File: server/services/v1/realtime/serialization.go [hunk 2]
```diff
@@ -80,64 +82,67 @@ func EncodeAsJSON(_ api.EventType, event proto.Message) ([]byte, error) {
...
+
+	if err != nil {
+		return nil, err
 	}
-	return nil, fmt.Errorf("unsupported event '%d'", encodingType)
+	return req2.(*api.RealTimeMessage), nil
 }
 
 func DecodeEvent(encodingType WSEncodingType, eventType api.EventType, message []byte) (proto.Message, error) {
+	switch eventType {
```
Comment (efirs) [inline server/services/v1/realtime/serialization.go:104]: We could simplify this:
```
var event interface{}

switch eventType {
case api.EventType_auth:
		event = &api.AuthEvent{}
case ...

}

return decode(encodingType, event, message)
```

File: server/services/v1/realtime/serialization.go [hunk 3]
```diff
@@ -80,64 +82,67 @@ func EncodeAsJSON(_ api.EventType, event proto.Message) ([]byte, error) {
 
 func EncodeAsMsgPack(eventType api.EventType, event proto.Message) ([]byte, error) {
 	var buf bytes.Buffer
-	if err := buf.WriteByte(byte(eventType)); err != nil {
```
Comment (himank) [inline server/services/v1/realtime/serialization.go:83]: If you remove it then how we will be decoding the events?

File: server/services/v1/realtime/device_session.go (hunk 1 above)
Comment (himank) [reply server/services/v1/realtime/device_session.go:150]: We can log it as an error, this should be rare.

File: server/services/v1/realtime/serialization.go (hunk 2 above)
Comment (garrensmith) [reply server/services/v1/realtime/serialization.go:109]: Thanks, I've improved it based on that.

**Comment by garrensmith** [comment]
Running `make test` passes locally:

```
Attaching to tigris_test
tigris_test  | make -C api/proto generate GEN_DIR=../../api/server/v1 API_DIR=..
tigris_test  | make[1]: Entering directory '/go/src/tigris/api/proto'
tigris_test  | make[1]: Nothing to be done for 'generate'.
tigris_test  | make[1]: Leaving directory '/go/src/tigris/api/proto'
tigris_test  | yq --exit-status 'tag == "!!map" or tag== "!!seq"' .github/workflows/*.yaml config/*.yaml >/dev/null
tigris_test  | shellcheck scripts/*
tigris_test  | shellcheck test/docker/grafana/*
tigris_test  | golangci-lint --timeout=5m run --fix
tigris_test  | go test -cover -race -tags=test,integration  ./...
tigris_test  | ok  	github.com/tigrisdata/tigris/api/server/v1	(cached)	coverage: 1.2% of statements
tigris_test  | ?   	github.com/tigrisdata/tigris/cmd/consistency	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/cmd/consistency/workload	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/cmd/realtime	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/cmd/realtime/app	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/errors	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/internal	(cached)	coverage: 13.4% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/keys	(cached)	coverage: 21.4% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/lib/container	(cached)	coverage: 57.6% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/lib/date	(cached)	coverage: 100.0% of statements
tigris_test  | ?   	github.com/tigrisdata/tigris/lib/json	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/lib/uuid	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/query/aggregation	(cached)	coverage: 61.2% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/query/expression	(cached)	coverage: 76.7% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/query/filter	(cached)	coverage: 62.4% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/query/read	(cached)	coverage: 26.7% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/query/search	(cached)	coverage: 31.5% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/query/sort	(cached)	coverage: 91.7% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/query/update	(cached)	coverage: 84.4% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/schema	(cached)	coverage: 76.0% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/schema/lang	(cached)	coverage: 78.0% of statements
tigris_test  | ?   	github.com/tigrisdata/tigris/server	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/server/cdc	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/server/config	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/server/defaults	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/server/metadata	(cached)	coverage: 48.1% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/server/metrics	(cached)	coverage: 62.6% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/server/middleware	(cached)	coverage: 9.9% of statements
tigris_test  | ?   	github.com/tigrisdata/tigris/server/muxer	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/server/quota	(cached)	coverage: 88.2% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/server/request	(cached)	coverage: 17.2% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/server/search	(cached)	coverage: 80.5% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/server/services/v1	(cached)	coverage: 0.4% of statements
tigris_test  | ?   	github.com/tigrisdata/tigris/server/services/v1/auth	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/server/services/v1/database	(cached)	coverage: 14.8% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/server/services/v1/realtime	(cached)	coverage: 20.4% of statements
tigris_test  | ?   	github.com/tigrisdata/tigris/server/tracing	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/server/transaction	(cached)	coverage: 1.1% of statements
tigris_test  | ?   	github.com/tigrisdata/tigris/server/types	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/store/cache	(cached)	coverage: 53.9% of statements
tigris_test  | ok  	github.com/tigrisdata/tigris/store/kv	(cached)	coverage: 38.0% of statements
tigris_test  | ?   	github.com/tigrisdata/tigris/store/search	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/templates	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/test/config	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/test/v1/client	(cached)	coverage: [no statements]
tigris_test  | ok  	github.com/tigrisdata/tigris/test/v1/server	(cached)	coverage: [no statements]
tigris_test  | ?   	github.com/tigrisdata/tigris/util	[no test files]
tigris_test  | ?   	github.com/tigrisdata/tigris/util/log	[no test files]
tigris_test  | ok  	github.com/tigrisdata/tigris/value	(cached)	coverage: 51.7% of statements
```

I'm not quite sure why CI fails.

**Comment by garrensmith** [comment]
Tests passed after rerunning them a few times. They could be a slightly flaky

File: test/docker/docker-compose.yml [hunk 4]
```diff
@@ -101,6 +100,7 @@ services:
     depends_on:
       - tigris_server
       - tigris_db2
+      - tigris_cache
```
Comment (himank) [inline test/docker/docker-compose.yml:103]: is this causing any issues for you? Shouldn't be needed as server2 is dependent on the server which waits for the cache.

**Review by himank — APPROVED** [review]

**Review by himank — APPROVED** [review]

**Pull request merged**